### PR TITLE
feat(gcfv2): migrate OCR to Cloud Functions v2

### PR DESCRIPTION
**Pull request by WildSunLove**

### CODE REVIEW

File: functions/v2/ocr/app/index.js [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+// [START functions_ocr_save]
+/**
+ * This function is exported by index.js, and is executed when
+ * a message is published to the Cloud Pub/Sub topic specified
+ * by the RESULT_TOPIC environment variable. The function saves
+ * the data packet to a file in GCS.
+ *
+ * @param {object} cloudEvent The CloudEvent containing the Pub/Sub Message object.
+ */
+functions.cloudEvent('saveResult', async cloudEvent => {
```
Comment (ace-n) [inline functions/v2/ocr/app/index.js:190]: Should each function be in its own directory?

(For one thing, I think each function has a different set of dependencies.)

File: functions/v2/ocr/app/test/index.test.js [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+          name: filename,
+        },
+      });
+
+      const server = getTestServer('processImage');
+      await supertest(server).post('/').send(cloudEvent);
+      assert.ok(
+        stubCons.log.calledWith(`Looking for text in image ${filename}`)
+      );
+      assert.ok(stubCons.log.calledWith('Extracted text from image:', text));
```
Comment (ace-n) [inline functions/v2/ocr/app/test/index.test.js:86]: Nit: instead of `stubCons`, can we use the global `console` object here?

**Review by ace-n — CHANGES_REQUESTED** [review]

File: functions/v2/ocr/app/test/index.test.js (hunk 2 above)
Comment (WildSunLove) [reply functions/v2/ocr/app/test/index.test.js:86]: Oh woops yes that was me testing something out and forgot to remove it.

File: functions/v2/ocr/app/index.js (hunk 1 above)
Comment (WildSunLove) [reply functions/v2/ocr/app/index.js:190]: Good question, I'm not sure, but I'll look into it.

Comment (WildSunLove) [reply functions/v2/ocr/app/index.js:190]: Also worth noting most of this code is not new, it's just slightly modified from here:
https://github.com/GoogleCloudPlatform/nodejs-docs-samples/tree/main/functions/ocr

But if you think there's changes worth making here, I'm open to doing so.

Comment (ace-n) [reply functions/v2/ocr/app/index.js:190]: Clarified this internally

**TL;DR:** this is fine as-is. 🙂

File: functions/v2/ocr/app/index.js [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+'use strict';
+
+// [START functions_ocr_setup]
+// Get a reference to the Pub/Sub component
+const {PubSub} = require('@google-cloud/pubsub');
+const pubsub = new PubSub();
+// Get a reference to the Cloud Storage component
```
Comment (ace-n) [inline functions/v2/ocr/app/index.js:21]: **Nit:** add newline above this one

**Review by ace-n — APPROVED** [review]
LGTM, but it might be worth getting a second pair of eyes. (@muncus maybe?)

File: functions/v2/ocr/app/index.js [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+/**
+ * This function is exported by index.js, and is executed when
+ * a message is published to the Cloud Pub/Sub topic specified
+ * by the TRANSLATE_TOPIC environment variable. The function
+ * translates text using the Google Translate API.
+ *
+ * @param {object} cloudEvent The CloudEvent containing the Pub/Sub Message object
+ */
+functions.cloudEvent('translateText', async cloudEvent => {
+  const pubsubData = cloudEvent.data;
```
Comment (averikitsch) [inline functions/v2/ocr/app/index.js:146]: nit: might be nice to link to https://cloud.google.com/storage/docs/json_api/v1/objects

File: functions/v2/ocr/app/test/index.test.js [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+'use strict';
+
+require('..');
```
Comment (averikitsch) [inline functions/v2/ocr/app/test/index.test.js:17]: Is this needed? We should only require specifics.

File: functions/v2/ocr/app/index.js [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+  }
+  if (!name) {
+    throw new Error(
+      'Filename not provided. Make sure you have a "name" property in your request'
+    );
+  }
+
+  const results = await detectText(bucket, name);
+  console.log(`File ${name} processed.`);
+  return results;
```
Comment (averikitsch) [inline functions/v2/ocr/app/index.js:132]: CloudEvents functions don't return a value

**Review by averikitsch — CHANGES_REQUESTED** [review]

File: functions/v2/ocr/app/test/index.test.js (hunk 5 above)
Comment (WildSunLove) [reply functions/v2/ocr/app/test/index.test.js:17]: We need something along these lines just to evaluate the function calls. Previously we had `const program = require('..')` and then we called `program.translateText` and the others. But now we just need to have the index.js be evaluated.

File: functions/v2/ocr/app/index.js (hunk 6 above)
Comment (WildSunLove) [reply functions/v2/ocr/app/index.js:132]: Ah thanks, I think I had put this in there when I was testing something out after running into issues with testing against console log, but I figured that out. I'll remove this.

File: functions/v2/ocr/README.md [hunk 7]
```diff
@@ -0,0 +1,27 @@
+<img src="https://avatars2.githubusercontent.com/u/2810941?v=3&s=96" alt="Google Cloud Platform logo" title="Google Cloud Platform" align="right" height="96" width="96"/>
+
+# Google Cloud Functions - OCR (Optical Character Recognition) sample
+
+See:
+
+* [Cloud Functions OCR tutorial][tutorial]
+* [Cloud Functions OCR sample source code][code]
```
[the remark this answers is not in the document]
Comment (WildSunLove) [reply functions/v2/ocr/README.md:8]: Removed, thanks.

File: functions/v2/ocr/app/index.js (hunk 4 above)
Comment (WildSunLove) [reply functions/v2/ocr/app/index.js:146]: Added, thanks.

File: functions/v2/ocr/.gitignore [hunk 8]
```diff
@@ -0,0 +1,2 @@
+node_modules
+config.json
```
[the remark this answers is not in the document]
Comment (WildSunLove) [reply functions/v2/ocr/.gitignore:2]: Removed, thanks

File: functions/v2/ocr/app/index.js (hunk 6 above)
Comment (WildSunLove) [reply functions/v2/ocr/app/index.js]: Done

**Review by averikitsch — APPROVED** [review]
